Complete the Errors & Debugging section

toji opened the pull request:
Resolves the outstanding issues to define `push/popErrorScope()` and formalizes the definitions of error scopes and how errors get pushed onto them.


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/gpuweb/gpuweb/pull/2506.html" title="Last updated on Feb 11, 2022, 11:08 PM UTC (6edd8f9)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/gpuweb/gpuweb/2506/a4d154e...6edd8f9.html" title="Last updated on Feb 11, 2022, 11:08 PM UTC (6edd8f9)">Diff</a>

kainino0x commented on spec/index.bs:
```diff
@@ -9463,19 +9469,101 @@ partial interface GPUDevice {
+                        [=reject=] |promise| with an {{OperationError}} and stop.
+
+                        <div class=validusage>
+                            - |this| is not [=lose the device|lost=].
+                            - |this|.{{GPUDevice/[[errorScopeStack]]}}.[=list/size=] &gt; 0.
```
nit
```suggestion
                            - |this| must not be [=lose the device|lost=].
                            - |this|.{{GPUDevice/[[errorScopeStack]]}}.[=list/size=] must be &gt; 0.
```

kainino0x reviewed (approved):
I think we should have a bit of explanatory text on the "Telemetry" section describing the intended use of 'uncapturederror' (telemetry and otherwise unexpected errors, never for handling errors which should use error scopes instead). Also should figure out how to define an event type (`"uncapturederror"`) properly.

That can be later though. LGTM

toji commented:
Addressed feedback, and added a definition of the `uncapturederror` event. (Modeled after the [event definitions we did for WebXR](https://immersive-web.github.io/webxr/#event-types), which I think were modeled off of the even definitions for WebAudio?)

kainino0x commented on spec/index.bs:
```diff
@@ -9499,6 +9588,15 @@ partial interface GPUDevice {
 
 ## Telemetry ## {#telemetry}
 
+User agents must dispatch an <dfn event for="GPUDevice">uncapturederror</dfn> event on a
+{{GPUDevice}} when a {{GPUError}} is generated that is not obverved by any [=GPU error scope=]. The
```
```suggestion
User agents **may** dispatch an <dfn event for="GPUDevice">uncapturederror</dfn> event on a
{{GPUDevice}} when a {{GPUError}} is generated that is not observed by any [=GPU error scope=]. The
```

kainino0x commented on spec/index.bs:
```diff
@@ -9499,6 +9588,15 @@ partial interface GPUDevice {
+event must be of type {{GPUUncapturedErrorEvent}}.
+
+Note: {{GPUDevice/uncapturederror}} events are intended to be used for telemetry and reporting
+unexpected errors, and should not be used for handling known error cases that may occur during
```
```suggestion
unexpected errors. They may not be dispatched for all uncaptured errors (for example, there may be a limit on the number of errors surfaced), and should not be used for handling known error cases that may occur during
```

kainino0x commented on spec/index.bs:
```diff
@@ -9463,19 +9469,102 @@ partial interface GPUDevice {
+            1. Set |scope| to the previous [=list/item=] of
+                |device|.{{GPUDevice/[[errorScopeStack]]}}.
+
+        1. [=Queue a task=] to fire an {{GPUUncapturedErrorEvent}} named
```
I forgot, this needs to be optional.

toji replied:
Done.

toji commented on spec/index.bs:
```diff
@@ -9463,19 +9469,102 @@ partial interface GPUDevice {
+    validation error</dfn> for {{GPUDevice}} |device|, run the following steps:
+
+    1. Let |error| be a new {{GPUValidationError}} with an appropriate error message.
+    1. Push |error| [$to the current error scope$] of |device|.
```
Yeah, that does feel weird, huh? Refactored to use something much closer to your suggestion.
